### PR TITLE
Java support

### PR DESCRIPTION
**Pull request by Phat3**
Add support for Java bytecode analysis

### CODE REVIEW

File: angr/__init__.py [hunk 1]
```diff
@@ -16,6 +16,9 @@
 Good luck!
 """)
 
+import warnings
+warnings.simplefilter('ignore')
```
Comment (ltfish) [inline angr/__init__.py]: Why?

Comment (Phat3) [reply angr/__init__.py]: I have a weird behaviour in this branch that I haven't figure out why it happens and how to solve it decently. The problem is the following: When I run the tests (or any other python program) with python inside the virtualenv created with this branch, the interpreter prints a lot of warnings (most of them are ResourceWarning) and it floods the output of the test cases. The weird thing is that those warnings are also present in angr-master (if printed explicitly) but they are in some way hidden.

I tried to investigate changes in the environment between angr-master and angr-feat/mixed_java but the environment and the python packages installed are exactly the same. Launching the script with `python -W ignore` or setting the evironment variable `PYTHONWARNINGS="ignore"` does not solve the problem.

The steps to reproduce the behavior are the following:

- `(angr-feat/mixed_java) python tests/test_0div.py ---> LOT OF WARNINGS PRINTED` 

- `(angr-master) python tests/test_0div.py ---> NO WARNING PRINTED` 

- `(angr-master) python -W always tests/test_0div.py ---> LOT OF WARNINGS PRINTED (same warnings printed in the branch feat/mixed_java)`

I put those two lines as a temporary fix to not flood your output during the execution of the test cases... I should have put a comment :sweat_smile:.

Have you ever encountered something similar?

Comment (ltfish) [reply angr/__init__.py]: No. It's possible that one of the other updated libraries are forcing all logging output to be printed out.

File: angr/analyses/cfg/cfg_base.py [hunk 2]
```diff
@@ -2120,6 +2136,25 @@ def _graph_traversal_handler(self, g, src, dst, data, blockaddr_to_function, kno
     # Other functions
     #
 
+    def _loc_to_funcloc(self, location):
+        """
+
+        :param int location:
+        :return:
+        """
+        return location
```
Comment (ltfish) [inline angr/analyses/cfg/cfg_base.py]: It seems to me that `_loc_to_funcloc()` is useless...

File: angr/analyses/cfg/cfg_fast.py [hunk 3]
```diff
@@ -1329,7 +1368,7 @@ def _job_key(self, job):
         return job.addr
 
     def _pre_analysis(self):
-
+
```
Comment (ltfish) [inline angr/analyses/cfg/cfg_fast.py]: Redundant spaces.

File: angr/analyses/cfg/cfg_fast.py [hunk 4]
```diff
@@ -1884,7 +1924,26 @@ def _scan_irsb(self, cfg_job, current_func_addr):
 
         return entries
 
+    def _create_entries_filter_target(self, target):
+        """
+
```
Comment (ltfish) [inline angr/analyses/cfg/cfg_fast.py]: Some comments on what this method does would be awesome!

File: angr/analyses/cfg/cfg_node.py [hunk 5]
```diff
@@ -1,12 +1,13 @@
...
 import logging
 
+from archinfo.arch_soot import SootAddressDescriptor
 import archinfo
 
 from ...codenode import BlockNode, HookNode
 from ...engines.successors import SimSuccessors
 
-_l = logging.getLogger(__name__)
+l = logging.getLogger(__name__)
```
Comment (ltfish) [inline angr/analyses/cfg/cfg_node.py]: Is the logger used anywhere in this file?

File: angr/block.py [hunk 6]
```diff
@@ -222,6 +222,38 @@ def instruction_addrs(self):
         return self._instruction_addrs
 
 
+class SootBlock(object):
```
Comment (ltfish) [inline angr/block.py]: `(object)` is unnecessary for Python 3.

File: angr/engines/soot/expressions/base.py [hunk 7]
```diff
@@ -0,0 +1,23 @@
+
+from . import translate_expr
+from ..values import translate_value
+
+
+class SimSootExpr(object):
```
Comment (ltfish) [inline angr/engines/soot/expressions/base.py]: `(object)` is not required in Python 3.

File: angr/engines/soot/statements/base.py [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+from archinfo.arch_soot import SootAddressDescriptor
+
+from ..exceptions import IncorrectLocationException
+from ..expressions import translate_expr
+from ..values import translate_value
+
+l = logging.getLogger('angr.engines.soot.statements.if')
+
+class SimSootStmt(object):
```
Comment (ltfish) [inline angr/engines/soot/statements/base.py]: `(object)` is not necessary for Python 3.

File: angr/knowledge_plugins/functions/function_manager.py [hunk 9]
```diff
@@ -7,6 +7,9 @@
 from ..plugin import KnowledgeBasePlugin
 
 from .function import Function
+from .soot_function import  SootFunction
```
Comment (ltfish) [inline angr/knowledge_plugins/functions/function_manager.py]: Use one space instead of two.

File: angr/knowledge_plugins/functions/function_manager.py [hunk 10]
```diff
@@ -7,6 +7,9 @@
...
+
+from archinfo.arch_soot import  SootMethodDescriptor
```
Comment (ltfish) [inline angr/knowledge_plugins/functions/function_manager.py]: Use one space instead of two.

File: angr/knowledge_plugins/functions/function_manager.py [hunk 11]
```diff
@@ -315,7 +327,7 @@ def function(self, addr=None, name=None, create=False, syscall=False, plt=None):
         :param bool create: Whether to create the function or not if the function does not exist.
         :param bool syscall: True to create the function as a syscall, False otherwise.
         :param bool or None plt: True to find the PLT stub, False to find a non-PLT stub, None to disable this
-                                 restriction.
+                                                         restriction.
```
Comment (ltfish) [inline angr/knowledge_plugins/functions/function_manager.py]: Why changing this line?

File: angr/knowledge_plugins/functions/soot_function.py [hunk 12]
```diff
@@ -0,0 +1,963 @@
...
+from itanium_demangler import parse
+
+import claripy
+from ...errors import SimEngineError, SimMemoryError
+from ...procedures import SIM_LIBRARIES
+
+l = logging.getLogger(name=__name__)
+
+
+class SootFunction(object):
```
Comment (ltfish) [inline angr/knowledge_plugins/functions/soot_function.py]: `(object)` is unnecessary in Python 3.

Comment (ltfish) [reply angr/knowledge_plugins/functions/soot_function.py]: Also, it seems to me that `SootFunction` is sharing *a lot of code* with `Function`. Maybe we should have a base class for those two classes?

File: angr/misc/reprs.py [hunk 13]
```diff
@@ -0,0 +1,14 @@
+
+from archinfo.arch_soot import SootAddressDescriptor, SootMethodDescriptor
+
+
+def repr_addr(addr, x=True):
+    if type(addr) in (int, long):
+        if x:
```
Comment (ltfish) [inline angr/misc/reprs.py]: This `if` check seems redundant to me.

File: angr/procedures/definitions/advapi32.py [hunk 14]
```diff
@@ -6,7 +6,7 @@
 
 lib = SimLibrary()
 lib.set_library_names('advapi32.dll')
-lib.add_all_from_dict(P['advapi32'])
+if 'advapi32' in P: lib.add_all_from_dict(P['advapi32'])
```
Comment (ltfish) [inline angr/procedures/definitions/advapi32.py]: Why this change?

File: angr/procedures/java_jni/class_and_interface_operations.py~e4a4af19e828c2d09b9ca32ce1829f06d1044695 [hunk 15]
```diff
@@ -0,0 +1,31 @@
+import logging
```
Comment (ltfish) [inline angr/procedures/java_jni/class_and_interface_operations.py~e4a4af19e828c2d09b9ca32ce1829f06d1044695]: This is a temporary file from some editor, right?

File: angr/procedures/java_jni/method_calls.py~e4a4af19e828c2d09b9ca32ce1829f06d1044695 [hunk 16]
```diff
@@ -0,0 +1,266 @@
+
```
Comment (ltfish) [inline angr/procedures/java_jni/method_calls.py~e4a4af19e828c2d09b9ca32ce1829f06d1044695]: This is a temporary file.

File: angr/procedures/java_jni/not_implemented.py~e4a4af19e828c2d09b9ca32ce1829f06d1044695 [hunk 17]
```diff
@@ -0,0 +1,26 @@
+import logging
```
Comment (ltfish) [inline angr/procedures/java_jni/not_implemented.py~e4a4af19e828c2d09b9ca32ce1829f06d1044695]: Temporary file again.

File: angr/procedures/java_jni/version_information.py~e4a4af19e828c2d09b9ca32ce1829f06d1044695 [hunk 18]
```diff
@@ -0,0 +1,12 @@
+
```
Comment (ltfish) [inline angr/procedures/java_jni/version_information.py~e4a4af19e828c2d09b9ca32ce1829f06d1044695]: Temporary file.

File: angr/project.py [hunk 19]
```diff
@@ -367,9 +375,9 @@ def _check_user_blacklists(self, f):
         parameters?
         """
         return not self._should_use_sim_procedures or \
-            f in self._exclude_sim_procedures_list or \
-            f in self._ignore_functions or \
-            (self._exclude_sim_procedures_func is not None and self._exclude_sim_procedures_func(f))
+               f in self._exclude_sim_procedures_list or \
+               f in self._ignore_functions or \
+               (self._exclude_sim_procedures_func is not None and self._exclude_sim_procedures_func(f))
```
Comment (ltfish) [inline angr/project.py]: Unnecessary changes.

File: angr/sim_state.py [hunk 20]
```diff
@@ -4,16 +4,19 @@
...
 
 import logging
-l = logging.getLogger(name=__name__)
 
 import angr # type annotations; pylint:disable=unused-import
 import claripy
 import ana
 import archinfo
 
+from archinfo.arch_soot import ArchSoot, SootAddressDescriptor
```
Comment (ltfish) [inline angr/sim_state.py]: Move this line upward by one line (it belongs to the above group).

File: angr/simos/javavm.py [hunk 21]
```diff
@@ -0,0 +1,433 @@
...
+                # multiarray = SimSootExpr_NewMultiArray.new_array(self.state, element_type, size)
+                # multiarray.add_default_value_generator(lambda s: SimSootExpr_NewMultiArray._generate_inner_array(s, element_type, sizes))
+                # return  multiarray
+            elif type_.endswith('[]'):
+                array = SimSootExpr_NewArray.new_array(state, type_[:-2], BVV(2, 32))
+                return array
+            else:
+                return SimSootValue_ThisRef.new_object(state, type_, symbolic=True, init_object=False)
+
+
```
Comment (ltfish) [inline angr/simos/javavm.py]: Too many blank lines.

File: angr/state_plugins/javavm_memory.py [hunk 22]
```diff
@@ -0,0 +1,391 @@
...
+from ..storage.memory import SimMemory
+from .keyvalue_memory import SimKeyValueMemory
+from .plugin import SimStatePlugin
+
+l = logging.getLogger(name=__name__)
+
+MAX_ARRAY_SIZE = 1000   # FIXME arbitrarily chosen limit
+
+
+
```
Comment (ltfish) [inline angr/state_plugins/javavm_memory.py]: Too many blank lines (two blank lines expected).

File: angr/storage/kvstore.py [hunk 23]
```diff
@@ -0,0 +1,41 @@
+
+from .memory import SimMemory
+
+
+class TypedVariable(object):
```
Comment (ltfish) [inline angr/storage/kvstore.py]: `(object)` is not necessary in Python 3.

File: angr/analyses/cfg/cfg_fast.py (hunk 3 above)
Comment (Phat3) [reply angr/analyses/cfg/cfg_fast.py]: Fixed now

File: angr/analyses/cfg/cfg_fast.py (hunk 4 above)
Comment (Phat3) [reply angr/analyses/cfg/cfg_fast.py]: This function slipped through during the rebase. It was an old function used in cfg_fast that is not used anymore. Removed now.

File: angr/analyses/cfg/cfg_node.py (hunk 5 above)
Comment (Phat3) [reply angr/analyses/cfg/cfg_node.py]: It is used in line `93`. I replaced it with `_l` as it was before.

File: angr/analyses/loop_analysis.py [hunk 24]
```diff
@@ -0,0 +1,355 @@
...
+from pysoot.sootir.soot_value import SootValue, SootLocal
+
+from ..block import SootBlockNode
+from ..errors import AngrLoopAnalysisError
+from . import register_analysis
+from .analysis import Analysis
+from .forward_analysis import ForwardAnalysis, LoopVisitor
+
+
+class VariableTypes(object):
```
[the remark this answers is not in the document]
Comment (Phat3) [reply angr/analyses/loop_analysis.py]: Fixed Now

File: angr/block.py (hunk 6 above)
Comment (Phat3) [reply angr/block.py]: Fixed Now

File: angr/engines/soot/expressions/base.py (hunk 7 above)
Comment (Phat3) [reply angr/engines/soot/expressions/base.py]: Fixed Now

File: angr/engines/soot/statements/base.py (hunk 8 above)
Comment (Phat3) [reply angr/engines/soot/statements/base.py]: Fixed Now

File: angr/knowledge_plugins/functions/function_manager.py (hunk 9 above)
Comment (Phat3) [reply angr/knowledge_plugins/functions/function_manager.py]: Fixed Now

File: angr/knowledge_plugins/functions/function_manager.py (hunk 10 above)
Comment (Phat3) [reply angr/knowledge_plugins/functions/function_manager.py]: Fixed Now

File: angr/knowledge_plugins/functions/function_manager.py (hunk 11 above)
Comment (Phat3) [reply angr/knowledge_plugins/functions/function_manager.py]: No reason as far as I can tell. I restore the previous change.

File: angr/misc/reprs.py (hunk 13 above)
Comment (Phat3) [reply angr/misc/reprs.py]: This file is an old file that slipped through during the poll request. It is not used anymore. Removed now.

File: angr/procedures/java_jni/class_and_interface_operations.py~e4a4af19e828c2d09b9ca32ce1829f06d1044695 (hunk 15 above)
Comment (Phat3) [reply angr/procedures/java_jni/class_and_interface_operations.py~e4a4af19e828c2d09b9ca32ce1829f06d1044695]: Yes correct, I didn't noticed these files :sweat_smile: . Removed now

File: angr/procedures/java_jni/method_calls.py~e4a4af19e828c2d09b9ca32ce1829f06d1044695 (hunk 16 above)
Comment (Phat3) [reply angr/procedures/java_jni/method_calls.py~e4a4af19e828c2d09b9ca32ce1829f06d1044695]: Yes correct, I didn't noticed these files :sweat_smile: . Removed now

File: angr/procedures/java_jni/not_implemented.py~e4a4af19e828c2d09b9ca32ce1829f06d1044695 (hunk 17 above)
Comment (Phat3) [reply angr/procedures/java_jni/not_implemented.py~e4a4af19e828c2d09b9ca32ce1829f06d1044695]: Yes correct, I didn't noticed these files :sweat_smile: . Removed now

File: angr/procedures/java_jni/version_information.py~e4a4af19e828c2d09b9ca32ce1829f06d1044695 (hunk 18 above)
Comment (Phat3) [reply angr/procedures/java_jni/version_information.py~e4a4af19e828c2d09b9ca32ce1829f06d1044695]: Yes correct, I didn't noticed these files :sweat_smile: . Removed now

File: angr/project.py (hunk 19 above)
Comment (Phat3) [reply angr/project.py]: Restored previous changes

File: angr/sim_state.py (hunk 20 above)
Comment (Phat3) [reply angr/sim_state.py]: Fixed Now

File: angr/simos/javavm.py (hunk 21 above)
Comment (Phat3) [reply angr/simos/javavm.py]: Fixed Now

File: angr/state_plugins/javavm_memory.py (hunk 22 above)
Comment (Phat3) [reply angr/state_plugins/javavm_memory.py]: Fixed Now

File: angr/storage/kvstore.py (hunk 23 above)
Comment (Phat3) [reply angr/storage/kvstore.py]: Fixed Now

File: angr/analyses/defuse_analysis.py [hunk 25]
```diff
@@ -0,0 +1,212 @@
...
+        for var, var_uses in state.uses.items():
+            if var not in s.uses:
+                s.uses[var] = var_uses.copy()
+            else:
+                s.uses[var] |= var_uses
+
+        return s
+
+
+class DefUseAnalysis(ForwardAnalysis, Analysis):
```
[the remark this answers is not in the document]
Comment (Phat3) [reply angr/analyses/defuse_analysis.py]: This class was implemented by you a long time ago when you started the Java support. Now it is not used anymore by anyone (as far as I can tell) and it has not been tested with the new codebase. Should I delete this file?

File: angr/block.py (hunk 6 above)
Comment (Phat3) [reply angr/block.py]: This class (as in the case of DefUseAnalysis) was implemented by you a long time ago when you started the Java support. Now it is not used anymore by anyone (as far as I can tell) and it has not been tested with the new codebase. Moreover, the other class that deals with loops (`LoopFinder`) already works with soot IR. I think we can remove this class

File: angr/knowledge_plugins/functions/soot_function.py (hunk 12 above)
Comment (Phat3) [reply angr/knowledge_plugins/functions/soot_function.py]: True. I think Andrea forgot to refactor that part. Refactored now.

File: angr/analyses/cfg/cfg_base.py (hunk 2 above)
Comment (Phat3) [reply angr/analyses/cfg/cfg_base.py]: True. As in other cases, this is an old function that has been introduced by error during the rebase. I cleaned it now

File: angr/analyses/defuse_analysis.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (ltfish) [reply angr/analyses/defuse_analysis.py]: Great. Now it's gone.

File: angr/__init__.py (hunk 1 above)
Comment (ltfish) [reply angr/__init__.py]: Now it's gone!

**Comment by rhelmot** [comment]
Suggestion: we switch to `'{}'.format()` style formatting because it lets us define per-class `__format__` methods, which lets us use hex formatting for both integers and soot descriptors.
